vpr/src/analytical_place/partial_legalizer.cpp

```diff
@@ -1587,72 +1669,158 @@ void BiPartitioningPartialLegalizer::partition_blocks_in_window(
+    lower_window_overfill.relu();
 
-    // Handle the unplaced blocks.
-    // To handle these blocks, we will try to balance the overfill in both
-    // windows. To do this we sort the unplaced blocks by largest mass to
-    // smallest mass. Then we place each block in the bin with the highest
-    // underfill.
-    // FIXME: I think large blocks (like carry chains) need to be handled special
-    //        early on. If they are put into a partition too late, they may have
-    //        to create overfill! Perhaps the partitions can hold two lists.
-    std::sort(unplaced_blocks.begin(),
-              unplaced_blocks.end(),
-              [&](APBlockId a, APBlockId b) {
-                  const auto& blk_a_mass = density_manager_->mass_calculator().get_block_mass(a);
-                  const auto& blk_b_mass = density_manager_->mass_calculator().get_block_mass(b);
-                  return blk_a_mass.manhattan_norm() > blk_b_mass.manhattan_norm();
-              });
-    for (APBlockId blk_id : unplaced_blocks) {
-        // Project the underfill from each window onto the mass. This gives us
-        // the overfill in the dimensions the mass cares about.
+    PrimitiveVector upper_window_utilization;
+    for (APBlockId blk_id : upper_contained_blocks) {
         const PrimitiveVector& blk_mass = density_manager_->mass_calculator().get_block_mass(blk_id);
-        PrimitiveVector projected_lower_window_underfill = lower_window_underfill;
-        projected_lower_window_underfill.project(blk_mass);
-        PrimitiveVector projected_upper_window_underfill = upper_window_underfill;
-        projected_upper_window_underfill.project(blk_mass);
-        // Put the block in the window with a higher underfill. This tries to
-        // balance the overfill as much as possible. This works even if the
-        // overfill becomes negative.
-        if (projected_lower_window_underfill.sum() >= projected_upper_window_underfill.sum()) {
-            lower_window.contained_blocks.push_back(blk_id);
-            lower_window_underfill -= blk_mass;
-        } else {
-            upper_window.contained_blocks.push_back(blk_id);
-            upper_window_underfill -= blk_mass;
-        }
+        upper_window_utilization += blk_mass;
+    }
+    PrimitiveVector upper_window_overfill = upper_window_utilization - upper_window_capacity;
+    upper_window_overfill.relu();
+
+    // If both windows are not overfilled, we are done.
+    if (lower_window_overfill.is_zero() && upper_window_overfill.is_zero()) {
+        lower_window.contained_blocks = std::move(lower_contained_blocks);
+        upper_window.contained_blocks = std::move(upper_contained_blocks);
+        return;
+    }
+
+    // 2) Resolve any overfill by moving blocks over the partition line.
+    // Here, we try to move blocks that are closest to the partition line such
+    // that the overfill on both sides is balanced.
+
+    // Try to move things from lower to upper greedily.
```

Comment:
The code for the lower and upper overfill and the block moving code for lower and upper are very similar, with just upper / lower transposed.  Could you refactor it to a helper routine or two that you pass references to the from list and to list t to and the list of blocks to move etc.?  And then just call the helper(s) twice with the appropriate references.